Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -275,13 +255,25 @@ size_t WaiterListManager::totalWaiterCount()
     return totalCount;
 }
 
-void WaiterListManager::cancelAsyncWaiter(const AbstractLocker& listLocker, Waiter* waiter)
+void Waiter::scheduleWorkAndClear(const AbstractLocker& listLocker, DeferredWorkTimer::Task&& task)
 {
-    ASSERT(waiter->isAsync());
-    RefPtr<DeferredWorkTimer::TicketData> ticket = waiter->ticket(listLocker);
-    if (ticket)
-        waiter->vm()->deferredWorkTimer->cancelPendingWork(ticket.get());
-    waiter->cancelTimer(listLocker);
+    ASSERT(m_isAsync && m_vm && !isOnList());
+    if (auto ticket = this->ticket(listLocker)) {
+        m_vm->deferredWorkTimer->scheduleWorkSoon(ticket.get(), WTFMove(task));
+        clearTicket(listLocker);
+    }
+    clearTimer(listLocker);
+}
+
+void Waiter::cancelAndClear(const AbstractLocker& listLocker)
+{
+    ASSERT(m_isAsync);
+    if (auto ticket = this->ticket(listLocker)) {
+        m_vm->deferredWorkTimer->cancelPendingWork(ticket.get());
+        m_vm->deferredWorkTimer->scheduleWorkSoon(ticket.get(), [](DeferredWorkTimer::Ticket) mutable { });
```

Comment:
Why does the lambda need to be mutable? Can you remind me why we need to schedule work when we clear? Why does cancelPendingWork not clear the ticket now that it can be `ThreadSafeWeakPtr`ed?

Reply:
> Why does the lambda need to be mutable?
No, we don't need it. I directly copied from the previous use of scheduling an empty task. Will change.

> Can you remind me why we need to schedule work when we clear?
`cancelPendingWork` just marks the TicketData in the canceled status. Then, `scheduleWorkSoon` will fire the work with the TicketData. Here, we just explicitly schedules an empty work after the cancelation.

> Why does cancelPendingWork not clear the ticket now that it can be ThreadSafeWeakPtred?
Yeah, maybe we can do that. Since TicketData `ThreadSafeWeakPtred`, we can definitely optimize the workflow in `DeferredWorkTimer`.

Reply:
https://bugs.webkit.org/show_bug.cgi?id=276538